database/json_file_storage.go

```diff
@@ -0,0 +1,175 @@
+	mutexes map[ids.ID]*sync.RWMutex
+	logger  logging.Logger
+}
+
+// NewJSONFileStorage creates a new JSONFileStorage instance
+func NewJSONFileStorage(logger logging.Logger, dir string, networks []ids.ID) (*JSONFileStorage, error) {
+	storage := &JSONFileStorage{
+		dir:     filepath.Clean(dir),
+		mutexes: make(map[ids.ID]*sync.RWMutex),
+		logger:  logger,
+	}
+
+	for _, network := range networks {
+		storage.mutexes[network] = &sync.RWMutex{}
+	}
+
+	_, err := os.Stat(dir)
+	if err == nil {
+		// dir already exist
+		// return the existing storage
+		return storage, nil
+	}
+
+	// 0755: The owner can read, write, execute.
+	// Everyone else can read and execute but not modify the file.
+	err = os.MkdirAll(dir, 0755)
+	if err != nil {
+		storage.logger.Error("failed to create directory",
+			zap.String("dir", dir),
+			zap.Error(err))
+		return nil, err
+	}
+
+	return storage, nil
+}
+
+// Get the latest chain state from the json database, and retrieve the value from the key
+func (s *JSONFileStorage) Get(chainID ids.ID, key []byte) ([]byte, error) {
+	currentState := make(chainState)
+	fileExists, err := s.read(chainID, &currentState)
+	if err != nil {
+		s.logger.Error(
+			"failed to read file",
+			zap.String("chainID", chainID.String()),
+			zap.Error(err),
+		)
+		return nil, err
+	}
+	if !fileExists {
+		return nil, errors.Errorf("file does not exist. chainID: %s", chainID)
+	}
+
+	if _, ok := currentState[string(key)]; !ok {
+		return nil, errors.Errorf("key does not exist. chainID: %s key: %s", chainID, key)
+	}
+
+	return []byte(currentState[string(key)]), nil
+}
+
+// Put the value into the json database. Read the current chain state and overwrite the key, if it exists
+func (s *JSONFileStorage) Put(chainID ids.ID, key []byte, value []byte) error {
+	currentState := make(chainState)
+	_, err := s.read(chainID, &currentState)
+	if err != nil {
+		s.logger.Error(
+			"failed to read file",
+			zap.String("chainID", chainID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	currentState[string(key)] = string(value)
+
+	return s.write(chainID, currentState)
+}
+
+// write value into the file
+func (s *JSONFileStorage) write(network ids.ID, v interface{}) error {
+	mutex, ok := s.mutexes[network]
+	if !ok {
+		return errors.Errorf("network does not exist. chainID: %s", network.String())
+	}
+
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	fnlPath := filepath.Join(s.dir, network.String()+".json")
+	tmpPath := fnlPath + ".tmp"
+
+	b, err := json.MarshalIndent(v, "", "\t")
+	if err != nil {
+		return err
+	}
+
+	// write marshaled data to the temp file
+	// if write failed, the original file is not affected
+	// 0644 Only the owner can read and write.
+	// Everyone else can only read. No one can execute the file.
+	if err := os.WriteFile(tmpPath, b, 0644); err != nil {
+		return errors.Wrap(err, "failed to write file")
+	}
+
+	// move final file into place
+	if err := os.Rename(tmpPath, fnlPath); err != nil {
```

Comment:
Maybe for a future PR: If we want to keep a few old versions of the file, we can implement "log rotation" by renaming the existing file to a backup name (same name but with an added suffix) before overwriting it. Probably not necessary for our use case.